src/core/PwAuth.cpp

```diff
@@ -0,0 +1,193 @@
+                bytes[i / 2] = static_cast<unsigned char>(v);
+            } else
+                throw std::invalid_argument("invalid hex value");
+        } catch (std::invalid_argument const& e) {
+            Debug::log(ERR, "auth: invalid password_hash");
+            bytes = nullptr;
+        } catch (std::out_of_range const& e) {
+            // Should never happen, as 2-byte substrings should never go o-o-r.
+            Debug::log(CRIT, "auth: implementation error in hex2Bytes conversion");
+            bytes = nullptr;
+        }
+    }
+    return bytes;
+}
+
+static auto bytes2Hex(const unsigned char* bytes, std::size_t len) -> std::string {
+    std::stringstream ss;
+    ss << std::setw(2) << std::setfill('0') << std::hex;
+    for (std::size_t i = 0; i < len; ++i)
+         ss << (int)bytes[i];
+    return ss.str();
+}
+
+CPwAuth::CPwAuth() {
+
+    if (gcry_check_version(NEED_LIBGCRYPT_VERSION))
+        gcry_control(GCRYCTL_INITIALIZATION_FINISHED, 0);
+    else
+        Debug::log(CRIT, "libgcrypt too old");
+
+    if (gcry_control(GCRYCTL_INITIALIZATION_FINISHED_P)) {
+
+        // Handle the hash algorithm
+        static auto* const ALGO = (Hyprlang::STRING*)(g_pConfigManager->getValuePtr("general:hash_algorithm"));
+        static auto const  algn = std::string(*ALGO);
+        m_iAlgo                 = gcry_md_map_name(algn.c_str());
+        m_iDigestLen            = gcry_md_get_algo_dlen(m_iAlgo);
+        if (m_iAlgo) {
+            static auto const err = gcry_err_code(gcry_md_test_algo(m_iAlgo));
+            if (err == GPG_ERR_NO_ERROR) {
+
+                // Handle the salt
+                static auto* const SALT = (Hyprlang::STRING*)(g_pConfigManager->getValuePtr("general:hash_salt"));
+                m_sSalt                 = std::string(*SALT);
+
+                // Handle the expected hash
+                static auto* const HASH = (Hyprlang::STRING*)(g_pConfigManager->getValuePtr("general:password_hash"));
+                static auto const  hash = std::string(*HASH);
+                if (hash.empty() || (hash.size() % 2) || (hash.length() != 2uL * m_iDigestLen)) {
+                    Debug::log(ERR, "auth: password_hash has incorrect length for algorithm {} (got: {}, expected: {})", algn, hash.size(), 2uL * m_iDigestLen);
+                    m_bLibFailed = true;
+                } else {
+                    m_sHash = hex2Bytes(hash);
+                    if (!m_sHash || hash.empty())
+                        m_bLibFailed = true;
+                }
+            } else {
+                // Might be due to FIPS mode
+                Debug::log(CRIT, "auth: hash algorithm unavailable: {}", algn);
+                m_bLibFailed = true;
+            }
+        } else {
+            Debug::log(ERR, "auth: unknown hash algorithm: {}", algn);
+            m_bLibFailed = true;
+        }
+    } else {
+        Debug::log(CRIT, "libgcrypt could not be initialized");
+        m_bLibFailed = true;
```

Comment:
Why not just crash? Maybe with a `RASSERT` or something. Without a crash you won't be able to log in, so I think that would make more sense.

Reply:
+1

Reply:
This happens during initialisation. You want the locker to lock up your system, because of an error that is outside of the user's realm of influence? Was thinking that the locker should rather not lock the system if it knows beforehand that it cannot reliably unlock afterwards.

In the end this is your decision to make. I just wanted to bring this up as a counter argument, as I have thought about this as well and came to the opposite conclusion.

Reply:
If we don't crash we lock up the system, but silently.
If we crash we lock up the system in a more obvious way.

So the decision is to make it crash :)

Reply:
The way I programmed it now is to not lock up the system at all but immediately return. Hyprlock has not yet told the Compositor to lock the system.

I make it crash if you want, but your statement about the current silent lock is not correct.

Reply:
From the perspective of the application, every case of `m_bLibFailed = true` looks identical. So, I would crash the application in all cases, then, not just when the library fails to initialise.